### PR TITLE
Updates recharts to v2.0.0 and fixes typing issues

### PR DESCRIPTION
**Pull request by leonardomaier**
## Hey, I just made a Pull Request!

Upgraded recharts to v2.0.0 and fixed all typing issues

Relates to #12609

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/cost-insights/src/components/CostOverviewCard/CostOverviewBreakdownChart.tsx [hunk 1]
```diff
@@ -186,10 +186,10 @@ export const CostOverviewBreakdownChart = ({
         ? DateTime.fromMillis(label)
         : DateTime.fromISO(label!);
     const dateTitle = date.toUTC().toFormat(DEFAULT_DATE_FORMAT);
-    const items = payload.map(p => ({
+    const items = payload.map((p, i) => ({
       label: p.dataKey as string,
-      value: formatGraphValue(p.value as number),
-      fill: p.fill!,
+      value: formatGraphValue(p.value as number, i),
+      fill: p.color!,
```
Comment (leonardomaier) [inline plugins/cost-insights/src/components/CostOverviewCard/CostOverviewBreakdownChart.tsx:192]: I'm double-checking if `p.color` is the right attribute here since `fill` was removed from the `Payload` type.

**Comment by freben** [comment]
You'll see that CI now says to move `@types/recharts` to devDependencies, which is a good thing!

File: plugins/cost-insights/api-report.md [hunk 2]
```diff
@@ -586,8 +586,8 @@ export interface ResourceData {
 // @public (undocumented)
 export type TooltipItem = {
   fill: string;
-  label: string;
-  value: string;
+  label: NameType | undefined;
+  value: ValueType;
```
Comment (freben) [inline plugins/cost-insights/api-report.md:590]: This seems like a widening of the possibilities, is that right? Could we have kept this as string + string only, and instead adapted the code underneath? I'm just thinking that maybe, widening the possibilities here could be an entirely different pull request if there turned out to be specifically a need for it.

File: plugins/cost-insights/api-report.md [hunk 3]
```diff
@@ -178,10 +178,10 @@ export type BarChartProps = {
...
   responsive?: boolean;
   displayAmount?: number;
-  options?: Partial<BarChartData>;
-  tooltip?: ContentRenderer<TooltipProps>;
-  onClick?: RechartsFunction;
-  onMouseMove?: RechartsFunction;
+  options?: Partial<BarChartOptions>;
+  tooltip?: ContentType<ValueType, NameType>;
+  onClick?: (...args: any[]) => void;
+  onMouseMove?: (...args: any[]) => void;
```
Comment (freben) [inline plugins/cost-insights/api-report.md:184]: Was there no more precise definition for these event handlers? Something that has the event object declared or so. If we promise absolutely _nothing_ for these parameters like this, then the type could just as well have been `() => void`

File: plugins/cost-insights/api-report.md [hunk 4]
```diff
@@ -178,10 +178,10 @@ export type BarChartProps = {
   resources: ResourceData[];
   responsive?: boolean;
   displayAmount?: number;
-  options?: Partial<BarChartData>;
-  tooltip?: ContentRenderer<TooltipProps>;
-  onClick?: RechartsFunction;
-  onMouseMove?: RechartsFunction;
+  options?: Partial<BarChartOptions>;
+  tooltip?: ContentType<ValueType, NameType>;
```
Comment (freben) [inline plugins/cost-insights/api-report.md:182]: Yeah here i'm wondering if this shouldn't just be a narrow, precise type like `{ title: string, description: string }` or whatever. Then that's the feature set we promise, for now, and extend it in the future if we ever deem it necessary.

The reason I'm bringing this up is, I'm essentially wondering if we can _remove_ as much explicit recharts-type-fiddling as possible. Look at it this way: if the type had been that in the first place, then the recharts update would not have had to make a breaking change to the public API surface here.

**Review by freben — COMMENTED** [review]
Alright, some initial thoughts

File: plugins/cost-insights/api-report.md (hunk 3 above)
Comment (leonardomaier) [reply plugins/cost-insights/api-report.md:184]: This was the definition of the `RechartsFunction` type on version v1.8.5:

```ts
export type RechartsFunction = (...args: any[]) => void;
```

Since the type was removed on v2.0.0, I just replaced it with its equivalence.

File: plugins/cost-insights/api-report.md (hunk 2 above)
Comment (leonardomaier) [reply plugins/cost-insights/api-report.md:590]: It makes sense, I'll undo these changes and keep them as a string

File: plugins/cost-insights/api-report.md (hunk 4 above)
Comment (leonardomaier) [reply plugins/cost-insights/api-report.md:182]: You're talking about the `tooltip?: ContentType<ValueType, NameType>`?

Since it's expecting a function that returns a `ReactNode` I've created a `TooltipRenderer` type like the following:

```ts
export type TooltipRenderer = ((props: TooltipProps<string, string>) => ReactNode)
```

We wouldn't depend on the `ContentType` type anymore, only `TooltipProps` which are the available props for the tooltip

File: .changeset/short-balloons-work.md [hunk 5]
```diff
@@ -0,0 +1,9 @@
+---
+'@backstage/plugin-bitrise': patch
+'@backstage/plugin-code-coverage': patch
+'@backstage/plugin-cost-insights': patch
```
[the remark this answers is not in the document]
Comment (leonardomaier) [reply .changeset/short-balloons-work.md:4]: You're right, I'll fix that

**Comment by leonardomaier** [comment]
@freben I just made some changes, let me know what you think 😄

File: plugins/cost-insights/src/components/CostOverviewCard/CostOverviewBreakdownChart.tsx [hunk 6]
```diff
@@ -168,28 +167,26 @@ export const CostOverviewBreakdownChart = ({
...
       typeof label === 'number'
         ? DateTime.fromMillis(label)
         : DateTime.fromISO(label!);
     const dateTitle = date.toUTC().toFormat(DEFAULT_DATE_FORMAT);
-    const items = payload.map(p => ({
+    const items = payload.map((p, i) => ({
       label: p.dataKey as string,
-      value: formatGraphValue(p.value as number),
-      fill: p.fill!,
+      value: formatGraphValue(p.value as unknown as number, i),
```
Comment (freben) [inline plugins/cost-insights/src/components/CostOverviewCard/CostOverviewBreakdownChart.tsx:188]: this cast looks scary :D could possibly have been `formatGraphValue(Number(p.value), i)`

**Review by freben — COMMENTED** [review]
Alright, this is starting to come together. I see there's some CI error still.

**Comment by leonardomaier** [comment]
So, it looks like the `BuildTimeline.test.tsx` tests are broken. 

I took a look, and it seems to be caused by a recharts lib dependency called `react-resize-detector` which was on version v2.3.0 before the upgrade. After upgrading recharts, it goes to version 7.1.2 which is a huge jump.

I found out this [already happened with other people](https://github.com/maslianok/react-resize-detector/issues/217), and a code snippet was added to the documentation as you can see here:

https://github.com/maslianok/react-resize-detector#testing-with-enzyme-and-jest

After adding the code snippet to the test suite, it worked fine but I got some TS errors:

image

Not sure if there is any solution to that besides ignoring the error, but I'll dig into it.

**Comment by leonardomaier** [comment]
Found an interesting discussion related to the CI error:

https://github.com/recharts/recharts/commit/3115b4d869042f7daf94a267117f75adc5049e23

Still trying out some approaches.

**Comment by leonardomaier** [comment]
Upgrading recharts lib to v2.0.0 brought problems related to its dependencies. For example, recharts use [d3-*](https://github.com/d3/d3-array) (array, shape, interpolation, etc...) as a dependency and at some point, `d3` files were published as modules that broke the jest transform pipeline.

I ran all tests locally and the following error was thrown:

```
.../node_modules/d3-array/src/index.js:1
({"Object.<anonymous>":function(module,exports,require,__dirname,__filename,jest){export * from "d3-array";
SyntaxError: Unexpected token 'export'
```

In some way, jest wasn't able to resolve the dependencies from d3, so I followed what they did [in this issue thread](https://github.com/facebook/jest/issues/12036) which is a very similar error. I've added all d3 libs in the `moduleNameMapper` in `jest.js` and it worked fine.

Besides that, I added the `ResizeObserver` in some additional component tests, related to `react-resize-detector` upgrade like [I mentioned here](https://github.com/backstage/backstage/pull/13517#issuecomment-1238064316).

Btw, all checks went green. Finally.

File: packages/cli/config/jest.js [hunk 7]
```diff
@@ -127,6 +127,7 @@ async function getProjectConfig(targetPath, displayName) {
     collectCoverageFrom: ['**/*.{js,jsx,ts,tsx,mjs,cjs}', '!**/*.d.ts'],
     moduleNameMapper: {
       '\\.(css|less|scss|sss|styl)$': require.resolve('jest-css-modules'),
+      '^d3-(.*)$': 'd3-$1/dist/d3-$1',
```
Comment (freben) [inline packages/cli/config/jest.js:130]: Sorry but this won't work well in practice. If we need to have overrides for individual packages it'll be very hard to maintain (depending on what version of d3 you happen to be on, _transitively_, things may break spectacularly or not, and it'll also depend on what version of the CLI you'd be on at the time).

We'll have to find a way to make this change unnecessary I'm afraid. Not having fully grasped the underlying problem yet, I don't have any great recommendations at this point myself.

**Review by freben — CHANGES_REQUESTED** [review]
Alright, highly appreciated work on this! However, the CLI change will have to be reverted; hopefully there's some other way forward for that one.

**Comment by freben** [comment]
I just tried to push a change without the mapping, just to see what CI does of it

**Comment by leonardomaier** [comment]
Going to dig into it a little bit more and search for alternatives.

**Comment by leonardomaier** [comment]
@freben I've made some changes:

- To solve issues related to `d3-*` libs, I removed the parser config for `tsx` files on `jest.js` (based [on this thread](https://github.com/swc-project/jest/issues/98))
- After the change mentioned above, the `LogoIcon.jsx` component started to fail, and after some research, I realized that we don't use the `jsx` extension for our components, so I changed it to `tsx`. Not sure if it was defined that way by mistake, but I fixed it.
- `BuildTimeline` component tests were failing due to the `ResponsiveContainer` component from `recharts`, so I needed to mock it and it solved the problem.

Besides that, you can see that a check failed, but it doesn't seem to be related to these changes. Let me know what you think.

**Comment by freben** [comment]
Note that the top of the LogoIcon file says

```ts
// @ts-check
// NOTE: This file is intentionally .jsx, so that there is one file in this repo where we make sure .jsx files work.
```

So that one is really meant to be jsx so that we can check that those flows work too...

**Comment by leonardomaier** [comment]
@freben you're right, I'll undo this change.

I've found an interesting issue on the SWC repository that was tagged as a bug:

https://github.com/swc-project/swc/issues/4911

The error above is what happens if I keep the parser for tsx files

**Comment by freben** [comment]
Ah yeah great find! The playground link in the repo concurs with you ... maybe we can put this one on ice for a little bit and hope that an swc update fixes things for us soon. Annoying that you hit these roadblocks, but we're grateful for the thoroughness and investigations around it!

**Comment by leonardomaier** [comment]
Referencing the PR that might unblock this: https://github.com/swc-project/swc/pull/6067

**Comment by leonardomaier** [comment]
Looks like the fix is planned to be released in swc@1.3.9

**Comment by leonardomaier** [comment]
@freben upgraded to `swc@1.3.9` and it fixed the parsing issue. All checks have passed, by the way.

File: packages/cli/package.json [hunk 8]
```diff
@@ -52,7 +52,7 @@
     "@svgr/plugin-svgo": "6.3.x",
     "@svgr/rollup": "6.3.x",
     "@svgr/webpack": "6.3.x",
-    "@swc/core": "^1.2.239",
+    "@swc/core": "^1.3.9",
```
Comment (freben) [inline packages/cli/package.json:55]: Just one final thing as a result of this.

- You need to make a separate changeset for the cli as well, now. Patch level is fine.
- Can you change the same thing in `storybook/package.json` too? That doesn't need a changeset.
- After this change, users will need to be on a sufficiently recent CLI to be able to build. I wonder if there will be any complex effects of that. I think it would be good if you could make a completely separate changeset for `@backstage/create-app`, at a patch level, explicitly pointing out that we have updated to a bug fixed version of SWC and that you may want to run `yarn backstage-cli versions:bump` to get on the latest version including the CLI itself. The changeset for that package is looked at extra carefully by a lot of people.

**Review by freben — COMMENTED** [review]
Almost done :) Code is fine!

File: .changeset/selfish-kiwis-matter.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/create-app': patch
+---
+
+Updated swc to v1.3.9 which fixes a `.tsx` parser bug. You may want to run `yarn backstage-cli versions:bump` to get on latest version including the CLI itself.
```
Comment (freben) [inline .changeset/selfish-kiwis-matter.md:5]: you can just put swc in backticks and then it'll stop complaining

**Review by freben — APPROVED** [review]
Alright, let's just get it to go green and then we're good to go!

**Comment by suuus** [comment]
Thank you for contributing during Hacktoberfest 🍂! You can claim the Backstage Hacktoberfest Holopin at https://bck.st/hacktoberfest-holopin 🙌🏻
